R/resample_draws.R

```diff
@@ -0,0 +1,176 @@
+#' Resample Draws Objects
+#'
+#' Resample [`draws`] objects according to given weights obtained,
+#' for instance, through importance sampling.
+#'
+#' @template args-methods-x
+#' @param weights A vector of positive weights of length equal to the
+#' number of draws in `draws` (as obtained by method [`ndraws`]).
+#' Draws with higher weights will be selected more often, usually
+#' roughly proportional to their weights. How exactly the weights are handled
+#' depends on the `method` argument. Weights will be internally normalized
+#' if necessary.
+#' @param method Name of the resampling method being applied.
+#' Possible choices are `"simple"` for simple random resampling with
+#' replacement, `"simple_no_replace"` for simple random resampling without
+#' replacement, `"stratified"` for stratified resampling with replacement
+#' and `"deterministic"` for deterministic resampling with replacement.
+#' Currently, `"stratified"` is the default as it has comparably
+#' low variance and bias with respect to ideal resampling. The latter would
+#' sample perfectly proportional to the weights but is not possible in practice
+#' where only a finite number of draws are available. For more details
+#' about resampling methods, see Kitagawa (1996).
+#' @param ndraws Number of draws to be returned. By default (`NULL`),
+#' `ndraws` is set internally to the total number of draws in `x` if sensible.
+#' @template args-methods-dots
+#' @template return-draws
+#'
+#' @details It is only possible to resample `draws` formats in which draws
+#' can be subsetted directly without breaking the format's structure,
+#' that is, only in `draws_matrix` and `draws_df` objects. Upon usage of
+#' `resample_draws`, chain information will be silently dropped due to
+#' subsetting of individual draws (see [`subset_draws`] for details).
+#'
+#' @references
+#' Kitagawa, G., Monte Carlo Filter and Smoother for Non-Gaussian Nonlinear '
+#' State Space Models, *Journal of Computational and Graphical Statistics*,
+#' 5(1):1-25, 1996.
+#'
+#' @examples
+#' x <- as_draws_df(example_draws())
+#'
+#' # random weights for illustrationary purposes
+#' w <- runif(ndraws(x), 0, 10)
+#'
+#' # use default stratified sampling
+#' x_rs <- resample_draws(x, weights = w)
+#' summarise_draws(x_rs, default_summary_measures())
+#'
+#' # use simple random sampling
+#' x_rs <- resample_draws(x, weights = w, method = "simple")
+#' summarise_draws(x_rs, default_summary_measures())
+#'
+#' @export
+resample_draws <- function(x, ...) {
+  UseMethod("resample_draws")
+}
+
+#' @rdname resample_draws
+#' @export
+resample_draws.draws <- function(x, weights, method = "stratified",
+                                 ndraws = NULL, ...) {
+  if (!allow_resample_draws(x)) {
+    stop2("'resample_draws' requires formats which can be subsetted ",
```

Comment:
What is the reason for this?

Reply:
After subsetting draws, the structure will no longer be compatible with a draws_array or draws_list object since, after subsetting, different chains will have different lengths. Do you see a way around this?

Reply:
One option would be to drop the chain dimension (with a message). A similar thing technically happens `draws_df` objects already. I will work on this option.

Reply:
Ok. Should be working now for all formats.